File: server/generator.js
import seedrandom from "seedrandom";

const EXPLANATION =
  "Combinação formada por 2 números do Grupo A, 2 do Grupo B e 2 do Grupo C, priorizando o score do modelo com base em análise estatística como ferramenta auxiliar.";

export const createRng = (seed) => {
  if (seed === null || seed === undefined || seed === "") {
    return Math.random;
  }
  return seedrandom(String(seed));
};

export const generateCards = ({ scores, groups, rng, cardCount, combinationsPerCard }) => {
  if ((groups.A ?? []).length < 2 || (groups.B ?? []).length < 2 || (groups.C ?? []).length < 2) {
    throw new Error("Groups A, B, and C must have at least 2 numbers each.");
  }

  const cards = [];
  for (let i = 0; i < cardCount; i += 1) {
    const combinations = generateCombinations({ scores, groups, rng, combinationsPerCard });
    cards.push({ id: `card-${String(i + 1).padStart(3, "0")}`, combinations });
  }
  return cards;
};

const generateCombinations = ({ scores, groups, rng, combinationsPerCard }) => {
  const combinations = [];
  const seen = new Set();
  const maxAttempts = combinationsPerCard * 20;
  let attempts = 0;

  while (combinations.length < combinationsPerCard && attempts < maxAttempts) {
    attempts += 1;
    const numbers = [
      ...sampleGroup(groups.A, scores, rng, 2),
      ...sampleGroup(groups.B, scores, rng, 2),
      ...sampleGroup(groups.C, scores, rng, 2),
    ].sort((a, b) => a - b);

    const key = numbers.join("-");
    if (seen.has(key)) {
      continue;
    }
    seen.add(key);

    combinations.push({
      numbers,
      score: comboScore(numbers, scores),
      explanation: EXPLANATION,
    });
  }

  if (combinations.length < combinationsPerCard) {
    throw new Error("Could not generate enough unique combinations.");
  }

  return combinations;
};

const sampleGroup = (groupNumbers, scores, rng, count) => {
  const selected = [];
  let available = groupNumbers.map((number) => ({
    number,
    weight: scores[number] ?? 0,
  }));

  for (let i = 0; i < count; i += 1) {
    const totalWeight = available.reduce((sum, item) => sum + item.weight, 0);
    const useUniform = totalWeight <= 0;
    const threshold = rng() * (useUniform ? available.length : totalWeight);
    let cumulative = 0;
    let chosenIndex = 0;

    for (let idx = 0; idx < available.length; idx += 1) {
      cumulative += useUniform ? 1 : available[idx].weight;
      if (threshold <= cumulative) {
        chosenIndex = idx;
        break;
      }
    }

    const [chosen] = available.splice(chosenIndex, 1);
    selected.push(chosen.number);
  }

  return selected;
};

const comboScore = (numbers, scores) => {
  if (!numbers.length) return 0;
  const mean = numbers.reduce((sum, number) => sum + (scores[number] ?? 0), 0) / numbers.length;
  return Math.round(mean * 10000) / 100;
};

File: backend/app/services/generator.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Dict, List

import numpy as np

EXPLANATION_TEXT = (
    "Combinação formada por 2 números do Grupo A, 2 do Grupo B e 2 do Grupo C, "
    "priorizando o score do modelo com base em análise estatística como ferramenta auxiliar."
)


@dataclass(frozen=True)
class Combination:
    numbers: List[int]
    score: float
    explanation: str


@dataclass(frozen=True)
class Card:
    card_id: str
    combinations: List[Combination]


def generate_cards(
    scores: Dict[int, float],
    groups: Dict[str, List[int]],
    rng: np.random.Generator,
    card_count: int,
    combinations_per_card: int,
) -> List[Card]:
    if len(groups.get("A", [])) < 2 or len(groups.get("B", [])) < 2 or len(groups.get("C", [])) < 2:
        raise ValueError("Groups A, B, and C must have at least 2 numbers each.")

    cards: List[Card] = []
    for card_index in range(card_count):
        combinations = _generate_combinations_for_card(
            scores,
            groups,
            rng,
            combinations_per_card,
        )
        cards.append(
            Card(card_id=f"card-{card_index + 1:03d}", combinations=combinations)
        )

    return cards


def _generate_combinations_for_card(
    scores: Dict[int, float],
    groups: Dict[str, List[int]],
    rng: np.random.Generator,
    combinations_per_card: int,
) -> List[Combination]:
    combinations: List[Combination] = []
    seen: set[tuple[int, ...]] = set()
    max_attempts = combinations_per_card * 20
    attempts = 0

    while len(combinations) < combinations_per_card and attempts < max_attempts:
        attempts += 1
        numbers = _sample_group(groups["A"], scores, rng, 2)
        numbers += _sample_group(groups["B"], scores, rng, 2)
        numbers += _sample_group(groups["C"], scores, rng, 2)
        numbers = sorted(numbers)

        numbers_key = tuple(numbers)
        if numbers_key in seen:
            continue

        seen.add(numbers_key)
        combo_score = _combo_score(numbers, scores)
        combinations.append(
            Combination(numbers=numbers, score=combo_score, explanation=EXPLANATION_TEXT)
        )

    if len(combinations) < combinations_per_card:
        raise ValueError("Could not generate enough unique combinations.")

    return combinations


def _sample_group(
    group_numbers: List[int],
    scores: Dict[int, float],
    rng: np.random.Generator,
    count: int,
) -> List[int]:
    weights = np.array([scores.get(num, 0.0) for num in group_numbers], dtype=float)
    if weights.sum() <= 0:
        weights = np.ones_like(weights)
    probabilities = weights / weights.sum()

    selected_indices = rng.choice(len(group_numbers), size=count, replace=False, p=probabilities)
    return [group_numbers[i] for i in selected_indices]


def _combo_score(numbers: List[int], scores: Dict[int, float]) -> float:
    if not numbers:
        return 0.0
    mean_score = sum(scores.get(num, 0.0) for num in numbers) / len(numbers)
    return round(mean_score * 100, 2)
